Refactor depot management

lomamech opened the pull request:
- Adding property parent to depot
- Manage depots as a tree structure
- Using the tree structure to assign repositories to users
- Update migrate file

closes #4922

lomamech committed:
feature(Stock Requisition)
- Forbid the deletion of stock requisition if the requisition has been
  implied in any process in Client Side

lomamech committed:
Solve conflict with branch master

lomamech committed:
feature(Stock Requisition)
- Adding a link to depot movement receipt in case of depot movement from
  requisition

lomamech committed:
Movement fully completed: in case all inventories of the requisition have been received by the destination

lomamech committed:
feature(Stock Requisition)
- Adding a new component to select the requisition reference
- Addition of the link from the register of requisitions to stock
  movements

- Finalization of the management of the various requisition status

lomamech committed:
feature(Stock Requisition)
- Restrict the delivery of stocks by requisition to requesting departments
  and depots and to target depots

- Display requisition reference in stock exit to service
- Update and improve SQL request for getting balance for movement from
  requisition

lomamech committed:
Sanitaze code

lomamech committed:
Resolve conflit with branch Master

lomamech committed:
Solve conflict

lomamech committed:
Test: Update and fixe requisition integration test

lomamech committed:
Clean and sanitaze code

lomamech committed:
Test: Update data for test, define default depot

lomamech committed:
Test: Reset value

lomamech committed:
Test: Reset value stock lot

lomamech committed:
Test: stock lots in stock

lomamech committed:
Test: stock lots in stock reset data

lomamech committed:
Test: stock lots in stock reset data

lomamech committed:
refactor(Depot Management)
- Adding property parent to depot

lomamech committed:
Make property user in component DepotSelect Optional

lomamech committed:
Refactor: remove property depotUuid and serviceUuid from component bhRequisitionSelect by the using of requestorUuid

lomamech committed:
improvement(Stock Requisition)
- Define label for requisition select in $onInit
- Improve request for requisition riun both
- Use bhConstant for value of requisition complete status

lomamech committed:
Display list of items not available for output via a requisition

lomamech committed:
Removal of the restriction on the exit of stocks at the depots for the deposit defined to the user, for a possible update of the rights of the users at the depots

lomamech committed:
Use constant for define requisition status partial and excessive

lomamech committed:
test(Depot Manangement)
- Implemente integration test for Depot Management

lomamech committed:
test(Depot Manangement)
- Implemente integration test for Depot Management

lomamech committed:
test: refactor E2E test for Depot Management

lomamech committed:
test: refactor E2E test for Depot Management

mbayopanda commented on client/src/i18n/fr/depot.json:
```diff
@@ -19,6 +20,7 @@
     "NO_DEPOT" : "Aucun dépôt",
     "NO_DEPOT_ASSIGNED_TO_USER" : "Vous n'avez pas la permission d'acceder à un dépôt ou aucun dépôt n'est enregistrer jusque là, veillez contacter l'administrateur en cas de besoin",
     "NOT_ALLOWED_ACTIONS" : "Ce dépôt est protégé, vous ne pouvez effectuer des modifications sur ce dépôt",
+    "SELECT_MAIN_DEPOT" : "Sélectionner le dépôt principale",
```
`principal`

mbayopanda commented on client/src/modules/depots/depots.html:
```diff
@@ -29,7 +36,27 @@
+  <div class="container-fluid">
+    <div 
+      id="depot-grid"
+      ng-if="DepotCtrl.listStructure"
```
Is it necessary to have both grid ?

We could just keep the tree version of the grid

mbayopanda commented on client/src/modules/depots/depots.js:
```diff
@@ -90,6 +99,12 @@ function DepotManagementController(
 
+  function toggleTreeMode() {
+    vm.treeStructure = !vm.treeStructure;
+    vm.listStructure = !vm.listStructure;
```
Since `vm.listStructure` is the reverse of `vm.treeStructure` could we just have `vm.treeStructure` and handle its reverse as `!vm.treeStructure`

mbayopanda commented on server/models/schema.sql:
```diff
@@ -461,6 +461,7 @@ CREATE TABLE `depot` (
   `allow_exit_loss` TINYINT(1) UNSIGNED NOT NULL DEFAULT 0,
   `location_uuid` BINARY(16) NULL,
   `min_months_security_stock` SMALLINT(5) NOT NULL DEFAULT 2,
+  `parent` BINARY(16) NULL,
```
Is it possible to rename the column `parent_uuid` since its must store the uuid of the parent

mbayopanda commented on test/end-to-end/depots/depots.page.js:
```diff
@@ -39,6 +39,29 @@ class DepotPage {
+   * @param {boolean} hasLocation if true it will enable the option of adding location
+   * @param {array} location an array of location as [country_uuid, province_uuid, sector_uuid, village_uuid]
+   */
+  async createDepotByParent(item, hasWarehouse, hasLocation, location) {
```
```js
async createDepotByParent(item, isWarehouse, hasLocation, location) {
```

mbayopanda reviewed (commented):
The feature works well 👍 

I have just some minors comments

jmcameron commented on client/src/i18n/fr/depot.json:
```diff
@@ -1,6 +1,7 @@
 {
   "DEPOT" : {
     "ADD_DEPOT" : "Ajouter un dépôt",
+    "ADD_DEPENDANT_DEPOT": "Ajouter un dépôt dépendant",
```
Correct spelling is:  DEPENDENT   (not -ANT)

jmcameron reviewed (approved):
I tested its  behavior and it seems to work fine.  However, I found the user interface a little bit odd.   I do not think you need a "Select Main Depot" checkbox.   Just have a box that says "Parent Depot".  It would default to "None" but clicking in it would allow the selection of a parent:

![image](https://user-images.githubusercontent.com/62145/94035983-a7b85600-fd78-11ea-8834-fa566e04aaf8.png)

Also, while you're at it, I suggest changing the label "Join location" to "Set location" on the option above.   You will need to update the French label appropriately too.

lomamech committed:
Update depot property use parent_uuid instead parent

lomamech committed:
Fix and Resolve conflict

lomamech committed:
Fix and resolve conflict

lomamech committed:
Test: Implement E2E test for set Depot management rights to User

lomamech committed:
Test: Implement E2E test for set Depot management rights to User

lomamech committed:
Merge conflict with Branch Master

lomamech committed:
Resolve conflict with branch master

lomamech committed:
Resolve conflict with branch master

lomamech committed:
Fix and resolve conflict with branch master

lomamech committed:
Fix and resolve conflict with branch master

lomamech committed:
Fix and resolve conflict with branch master

lomamech committed:
Fix and resolve conflict with branch master

lomamech committed:
Clean and sanitaze code

lomamech committed:
Clean and sanitaze code

lomamech committed:
Fix and clean rebasing with branch master, set loading in Assignation
depots to user

lomamech committed:
Update of the repository management interface, removal of the change in display mode by removing the list mode

lomamech committed:
Clean and sanitaze comment

lomamech committed:
Clean and sanitaze comment

jmcameron reviewed (approved):
I tested this and was pleased with the way it works now.    Thanks for making the changes.   I ran into one small issue:  When you have a Depot with a parent and you edit it, if you do the "Clear" for it, nothing happens.  I think it should remove the parent.  Then clicking the [Submit] button would save that change.

Also, a cosmetic suggestion:  I suggest (if possible) to make tree default to all branches/leaves open, not just showing the top level.   The user could always collapse the branches as desired.  I was not 100% sure about this suggestion.  Do most possible installations deal with a lot of depots?   If the answer is yes, then maybe the way it is now would be best.  

Finally, other than the minor issue with the "Clear" button, I think this is ready to go.  However, you may want to check with @jniles about whether to hold off on releasing this until after the major release that he is working on now.

mbayopanda reviewed (approved):
The feature is working fine.

I have just noticed that building tree on the client with the limit filter has some bad user experience, if the limited loaded depots doesn't have root parent they will not be displayed in the depot registry.

jniles committed:
fix(vouchers): document_uuid on the payment voucher

Adds the document_uuid of the payment to the payment voucher.

Fixes a bug in production.

lomamech committed:
Improvement: Expend all rows in Tree mode for depot

lomamech committed:
Improvement: Expend all rows in Tree mode for depot

dependabot-preview[bot] committed:
chore(deps): bump tempy from 0.7.0 to 0.7.1

Bumps [tempy](https://github.com/sindresorhus/tempy) from 0.7.0 to 0.7.1.
- [Release notes](https://github.com/sindresorhus/tempy/releases)
- [Commits](https://github.com/sindresorhus/tempy/compare/v0.7.0...v0.7.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

bors[bot] committed:
Merge #4958

4958: fix(vouchers): document_uuid on the payment voucher r=lomamech a=jniles

Adds the document_uuid of the payment to the payment voucher.

Fixes a bug in production.

Co-authored-by: Jonathan Niles <jonathanwniles@gmail.com>

lomamech committed:
test(Depot)
- Add test for clear parent depot, on depot management

lomamech committed:
test(Depot)
- Add test for clear parent depot, on depot management

bors[bot] committed:
Merge #4961

4961: chore(deps): bump tempy from 0.7.0 to 0.7.1 r=jniles a=dependabot-preview[bot]

Bumps [tempy](https://github.com/sindresorhus/tempy) from 0.7.0 to 0.7.1.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a href="https://github.com/sindresorhus/tempy/releases">tempy's releases</a>.</em></p>
<blockquote>
<h2>v0.7.1</h2>
<ul>
<li>Update dependencies  709bd3d</li>
</ul>
<p><a href="https://github.com/sindresorhus/tempy/compare/v0.7.0...v0.7.1">https://github.com/sindresorhus/tempy/compare/v0.7.0...v0.7.1</a></p>
</blockquote>


Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

jniles committed:
Release 1.15.0

bors[bot] committed:
Merge #4963

4963: Release 1.15.0 r=jniles a=jniles

🚀 Let's release bhima (currently at 1.14.0)

Changelog:
* fix(vouchers): document_uuid on the payment voucher (b83535262)
* fix: improve UI of stock exits (f65b7f191)
* fix(stock): do not show icon for stock out (e5e639637)
* Fix(unit list) put indicator report in dashboard folder (b86b0ba9c)
* feat(stock e2e tests) pull all tests in on block (6b16f7285)
* fix(survey): migrate survey form to ui-router (8cba6d791)
* fix(subsidies): migrate to latest ui-router (edfc01615)
* fix(services): migrate to new ui-router. (5d4d8b0ac)
* fix: payroll ui-state refs (f9946b4d3)
* fix: migrate indicator files to ui-router (fe7dd0fe9)
* fix(IPR): update tax routes to ui-router lastest (163565f2a)
* fix: more ui-router upgrades (cf47ccb08)
* fix: more upgrades to ui-router latest (9d5af5ce5)
* fix: more upgrades to ui-router (ad4abaf6f)
* fix: fill form controllers migrate ui-router (1df2b2dc5)
* fix: more ui-router migrations (2711aa966)
* fix(depots): upgrade ui-router for depots (5ef547b49)
* fix(data collection): migrate to ui-router (0b303c698)
* fix(config analysis): more ui-router upgrades (03842e071)
* fix(choice list mgmt): migrate to latest ui-router (79f7a8c59)
* fix(breakeven): fixes the breakeven for ui-router (2f834ddb8)
* deps: remove duplicate ui-router dependency (a9bbf090c)
* fix(accounts): update to latest ui-router (9bba0906b)
* fix(payroll): accounts config for ui-router (330eee5cd)
* fix(accounts): references proper use of $state (a7ec4f8d3)
* deps: upgrade ui-router to latest (5334d0adb)
* feat(inventory log) display correctly la chagement (19dd252bc)
* feat(posting journal) unpost transactions (7c272008c)
* fix: bump gemfile (ef1edfd02)
* get cmm from stock_movement_status (eb7e72602)
* fix(reports): stock expiration report (8135d8a5c)
* remove development env (7c6d5933b)
* Set the feature as optional in enterprise settings (a22e7b0ed)
* update e2e tests (dab512982)
* restrict access by users (8c954c33f)
* Handle depot permission (164fdbc62)
* feat: semaphore ci for integration tests (4f2204a32)
* fix(build): use .env for application config (eda2b5fa7)
* Hide 'Show items in stock' from action menu if none (2da7d0e8a)
* Update test (6d5960f9f)
* Update test (672696c04)
* Update test (2fc4c12a6)
* Update test (9536e1333)
* Update test (4d98d1fd9)
* feat(enterprise): add settings for supplier credit (89e50660d)
* fix(stock): credit supplier on entry from purchase (f936bd778)
* remove tracker in place keep importance (3760f8999)
* use of integer for importance levels (f3f3382c0)
* run mysql on 3306 (868f18b1f)
* Add tracker and importance (579b47099)
* add importance column (b013d90d4)
* Fixed test (0d80373ce)
* Copied in fixes from JNiles patch to my PR (4dc2eff81)
* Moved 'Include Exhausted Lots' into defaults tab for the lots registry. (00463e32a)
* test: force HTML rendering in travis w/ flag (fa29f6e5b)
* fix(receipts): allow stock movement lookups (c2e77dae3)
* fix(stock): movements should group by is_exit (a9045cc52)
* docs(stock): improve stock documentation (47a996df7)
* Test(Component Date Editor) - Addition of a new parameter to solve the problem of the date editor component for forms whose .header-image is not visible by adding a new component - Use bhDateEdit in Off Days E2E test (262a052cf)
* fix tooltip icon and color (9a35ce311)
* Use of tooltip and considering also warning about near to expire (99cb6752e)
* Fix server side and enhancements (9ed88a71e)
* Add search option for risky inventories (8cd539717)
* expiration alerts on the Article en stock registry (40cf91c05)
* lots alerts (19b8402dd)
* feat(CMM aggregate data calculation) set in_quantity and _out_quantity if start_date = end_date (f90bd4a6d)
* Fixed error handling (e45906fa5)
* fix(stock): skip tags query if empty result (e9dabb94e)
* bug(Stock exit from requisition) - fix error used the uuid of Depot From Data parameters (03b3432c3)
* fix(schema): FK constraints stock_movement_status (ac7ab4269)
* fix(adjustment): add inventory to data. (3c2d01410)
* perf(stock): use inventory for quantity calcs (0655e53e2)
* calling cmm data calculating for new stock movement (afd997c11)
* feat(stock): algorithm for quantity in stock (5ac6caffa)
* Implemented enterprise and depot-specific min months security stock parameter. (1ef2a225e)
* get CMM procedure (09fb0f469)
* calling cmm data calculating for new stock movement (4a314a561)
* fix(trial balance): catch title accounts w/ balances (1c48270f9)
* fix: ensure sane accounts in test data (fec2ad1a3)
* feat(stock): algorithm for quantity in stock (1a3e12f9f)
* Fixed typo on Exhausted lots (2a1d76433)
* Swapped inventory/lot labels on expiration warning block (00c3cfe5f)
* Improve stock lot highlighting in Stock Lot Registry (9a084541a)
* Use of GridRow (bd6138a01)
* Use of GridAction in place of GridRow (3c1954318)
* add tags (da50c8d66)
* Enhance tags queries (e7291f3cf)
* Apply suggestions from review (f1dbc2a78)
* skip buggy tests (857e8b3e8)
* update lots integration tests (f8b064a24)
* update tests (a99752a54)
* refactor tag modal for creation and update (91760aeef)
* fix and enhancements (347ff5b18)
* Implement search by tags (e6f92d09c)
* Tag features (9f31d344a)
* crud tags on lot (c9246b952)
* Add color in tags (18b917279)
* use location as default address (85da605e7)
* Fixed French translation, Connecter -> Se connecter (cb5c675ef)
* fix test data import (0674140c0)
* use of test db (1fecbb039)
* Update enterprise details on receipts (f5c7c3303)
* Add address in enterprise (9ae0f0eec)
* feat(Stock entry report)solve checking at least one issue (2b46b1549)
* test: update accents in inventory data (c7d67d2a1)
* fix(accounts): use correct $bhValue declaration (cf17761ac)
* feat(inventory): enforce account types on groups (8b3d60ec4)
* fix(stock): hide expiration info when irrelevant (f9dc44092)
* test for mysql function and procedures (b0b1469d0)
* feat(component) stock expired (ed790fc80)
* fix(lots): duration renders in the correct lang (c402e5bd2)
* docs(stock): add documenting comments to stock code (74b0d4816)
* Updated Connection/Connexion to Log in/Connected on main login page (ce81b70c1)
* fix collection capacity name to recovery capacity (95b446ff1)
* using async (ce0c2146b)
* test before updates (7a530927e)
* integration test (dd568044c)
* Fixe(Stock entry from donation) solve receipt rendering issue (8c1e95504)
* Updated end-to-end tests (38c8bd3e5)
* Update Stock-Exit to prevent expired lots/batches from being selected. Updated tests for bhima_test expiration date changes in PR 4812. (faa3deb17)
* patch for v1.14.0 about stock movement registry (0106bae46)
* add in risk of stock flag (dc2f9375a)
* feat(Stock) expiration report (f5f783960)
* Updated stock lot expiration dates some stock are not expired. Updated tests for changed lot expiration dates. (0f9b45608)
* fix(inventory): reverse COGS & asset accounts (f1551250c)
* feat(database schema) rename all foreign key constraints (ee513c2a1)
* fix: use correct CSP formulation in helmetjs (bcabc13d3)
* fix(stock): activate cancel button on depot modal (a59074d9a)
* fix(i18n): better translation on link to inventory (99917b980)
* feat(SQL procedures) drop columns and indexes if exist (c60467153)
* fix(csp): adds font and blob src to whitelist (629cae368)
* fix: ensure proper content-security policy. (841c39dc6)

Co-authored-by: Jonathan Niles <jonathanwniles@gmail.com>

dependabot-preview[bot] committed:
chore(deps-dev): bump eslint-plugin-import from 2.22.0 to 2.22.1

Bumps [eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import) from 2.22.0 to 2.22.1.
- [Release notes](https://github.com/benmosher/eslint-plugin-import/releases)
- [Changelog](https://github.com/benmosher/eslint-plugin-import/blob/master/CHANGELOG.md)
- [Commits](https://github.com/benmosher/eslint-plugin-import/compare/v2.22.0...v2.22.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

jniles committed:
chore: move migration script.

This commit resets the migration script that I forgot to do during the
release.

bors[bot] committed:
Merge #4966

4966: chore: move migration script. r=jmcameron a=jniles

This commit resets the migration script that I forgot to do during the release.

Co-authored-by: Jonathan Niles <jonathanwniles@gmail.com>

jniles committed:
docs: improve mysql documentation with docker

Improves the documentation for how to use mysql with docker.  Also adds
French sections addressing this.

jniles committed:
fix: allow multiple debtor invoice selections

This commit fixes the cash invoice selection modal by using $transition
instead of $state params.  For some reason, selecting the $state
multiple times does not work.

lomamech committed:
feature(Stock Requisition)
- Update label be explicit Partially delivered for requisition

lomamech committed:
Improve component bhDepotSelect rename params for select depot by User

lomamech committed:
improvement(Stock Requisition)
- Use ng-class instead bh-requistion-select
- Add property cacheable false in link from stock movement to stock
  requisition, and add filter allTime, and filter requisition by UUID

lomamech committed:
improvement(Stock Requisition)
- remplace icon for PDF in link from stock movement to requision by icon
  for file-text
- Remove not necessary link from inventory_type from checkExcessive balance

lomamech committed:
Improvement: Remove not necessary link from inventory_type from check
requition balance

bors[bot] committed:
Merge #4968

4968: fix: allow multiple debtor invoice selections r=mbayopanda a=jniles

This commit fixes the cash invoice selection modal by using `$transition` instead of `$state` params.  For some reason, selecting the `$state` multiple times does not work.

It's bug that is hitting our production systems.  We should do a quick check to make sure this isn't affecting more modals.  It's a tricky bug since it works the first time (and so, passes tests) but doesn't work after that.

Co-authored-by: Jonathan Niles <jonathanwniles@gmail.com>

lomamech committed:
Resolve conflict

lomamech committed:
Add property project_id in stock requisition

lomamech committed:
refactor(Stock Requisition)
- Add Projet abbr in document maps for requisition
- sanitaze code

lomamech committed:
improve(Triggers)
- Define reference for requisition by project

lomamech committed:
Finalize and Update Migrate file: featureStockRequisition

jniles committed:
fix(test): fix bh-date-editor tests

Cherry picked the bh-date-editor changes from #4837 and add the critical
await call to prevent test failures.

Closes #4919.

lomamech committed:
improvement(Stock Requisition)
- Refactor Stock exit from requisition add column stock_requisition_uuid
  in table stock_mouvement instead to use a join table

- Delete table stock_requisition_movement
- Refactor all link to filter requisition from stock_movement instead
  stock movement for requisition

lomamech committed:
Test: remove value

jmcameron committed:
Moved stock settings to dedicated page and table

jniles committed:
fix(stock): integration test errors

Make the integration tests pass by fixing a bug in the CMM calculation.

jniles committed:
fix(stock): security stock -> minimum stock

The security stock has become the minimum stock for a product.  Not sure
why.

bors[bot] committed:
Merge #4970

4970: fix(test): fix bh-date-editor tests r=lomamech a=jniles

Cherry picked the bh-date-editor changes from #4837 to bring them into `master` sooner.

Closes #4919.

Co-authored-by: Jonathan Niles <jonathanwniles@gmail.com>

lomamech committed:
Fix and resolve conclict

jmcameron committed:
Merge branch 'master' into stock_settings

jniles commented on client/src/modules/depots/templates/label.tmpl.html:
```diff
@@ -5,6 +5,6 @@
       { key : 'limit', value : 1000, cacheable: false }
     ]})" href>
-    <span>{{ row.entity.text }}</span>
+    <span style='padding-left : {{row.treeLevel * grid.appScope.indentTitleSpace}}px;' >{{ row.entity.text }}</span>
```
Could you use [`ngStyle`](https://docs.angularjs.org/api/ng/directive/ngStyle) here?

lomamech replied:
it generates error messages, maybe because there are several concatenations

jniles commented on client/src/i18n/en/depot.json:
```diff
@@ -1,6 +1,7 @@
 {
   "DEPOT" : {
     "ADD_DEPOT" : "Add a Depot",
+    "ADD_DEPENDENT_DEPOT": "Add dependent Depot",
```
A better phrase would be either "Add a sub depot" or "Add a child depot".

jniles commented on client/src/modules/depots/depots.js:
```diff
@@ -121,13 +133,23 @@ function DepotManagementController(
           return item;
         });
+
+        vm.gridOptions.data = FormatTreeData.formatStore(depotsData);
```
Cool 👍

jniles commented on client/src/modules/depots/templates/action.tmpl.html:
```diff
@@ -22,6 +22,13 @@
         <i class="fa fa-th-list"></i> <span translate>INVENTORY.VIEW_ARTICLES_IN_STOCK</span>
       </a>
     </li>
+    <li ng-if="row.entity.is_warehouse">
```
Aha, this is interesting.  Only warehouses can have child depots.

... that makes sense to me.  I'll be curious to see if that remains true throughout the project :).

jmcameron replied:
> Aha, this is interesting. Only warehouses can have child depots.

Is that true?  Don't even pharmacies in a hospital act as a small depot?  And a depot in a hospital would be the parent for that child pharmacy, but would not be a warehouse, right?

jniles commented on client/src/modules/users/UserDepotManagementModal.html:
```diff
@@ -19,14 +19,37 @@ <h4><i> {{ UsersDepotModalCtrl.user.display_name }} </i></h4>
+        </div>
+      </div>
+    </div>
+    <div ng-if="UsersDepotModalCtrl.loading" class="container text-center">
```
I don't think this should be `class="container"`.  Maybe you mean `row`?

bors[bot] committed:
Merge #4967

4967: docs: improve mysql documentation with docker r=jniles a=jniles

Improves the documentation for how to use mysql with docker.  Notably, I discuss how to ensure that you are running in the correct collation and with the mysql authentication plugin restored so node can connect to it.  Also adds a French section addressing this.





-----
[View rendered docs/en/for-developers/installing-bhima.md](https://github.com/jniles/bhima/blob/docs-improve-docker-configuration/docs/en/for-developers/installing-bhima.md)
[View rendered docs/fr/for-developers/installing-bhima.md](https://github.com/jniles/bhima/blob/docs-improve-docker-configuration/docs/fr/for-developers/installing-bhima.md)

Co-authored-by: Jonathan Niles <jonathanwniles@gmail.com>

lomamech committed:
Test: Update stock inventories

bors[bot] committed:
Merge #4921

4921: Feature stock requisition r=jniles a=lomamech

- Forbid the deletion of stock requisition if the requisition has been
  implied in any process in Client Side
- Adding a link to depot movement receipt in case of depot movement from
  requisition
-Movement fully completed: in case all inventories of the requisition have been received by the destination
- Adding a new component to select the requisition reference
- Addition of the link from the register of requisitions to stock
  movements
- Adding a new component to select the requisition reference
- Addition of the link from the register of requisitions to stock
  movements

- Finalization of the management of the various requisition status
- Restrict the delivery of stocks by requisition to requesting departments
  and depots and to target depots

- Display requisition reference in stock exit to service
- Update and improve SQL request for getting balance for movement from
  requisition
-Improved bh-depot-select component added filter option for user-defined
repositories
- Improvement of alert messages when out of stock by requisition
- Set data for build E2E test for exit stock by requisition
- Add requisition status excessive received
- Add stock_requisition uuid for document map
- Test: set data for E2E test and integration
- Test: Reset value for many E2E requisition test

closes #4849




Co-authored-by: Chris Lomame <lomamech@gmail.com>

jmcameron committed:
Merge branch 'master' into stock_settings

bors[bot] committed:
Merge #4946

4946: Stock settings page r=jniles a=jmcameron

Convert stock-related settings in Enterprise settings to their own page under the Stock menu.

The settings that were moved are:
1.  month_average_consumption
2.  default_min_months_security_stock
3.  enable_auto_purchase_order_confirmation
4.  enable_auto_stock_accounting
5.  enable_daily_consumption
6.  enable_strict_depot_permission
7.  enable_supplier_credit

**Suggested testing sequence**
- Use bhima_test (edit .env as needed)
- yarn build:db
- yarn dev
- Verify that there are no stock-related settings in the Enterprise settings
- Go to the "Stock Settings" under the stock menu
  - Change all the values then press [Submit].  Notice the success indicator on
    the top of the page
  - Do a full refresh of the page.  Verify that the values have not changed.
  - Change all the values then press [Submit]. 
  - Do a full refresh of the page.  Verify that the values have not changed.
  - Note the numeric values for month_average_consumption and
    default_min_months_security_stock (MMSS below)
- Go to Stock > Depot Management
  - Edit each depot and verify that MMSS is:
     - Depot Principal: 2
     - Depot Secondaire: 3
  - Start to create a depot
     - Notice that MMSS defaults to the last value set above
	 - Change MMSS value in Stock Settings and repeat above

The update also supports creating default stock settings if no entry exists
already.  This might occur in new installations.   Verify that it works like
this:
  - Notice defaults for stock settings in server/model/schema.sql
     - Look around line 1860 for: CREATE TABLE `stock_setting`
     - Note the default for enable_auto_stock_accounting in the SQL table is 1,
	but is 0 in the Bhima_test database.
  - Edit test/bhima.sql and temporarily disable: line 3000
     - Disable line: -- INSERT INTO `stock_setting` ...
  - Do 'yarn build:db'
    - Run mysql and verify that the stock_setting table is empty
  - Run 'yarn dev'
  - Refresh browser with BHIMA session.
  - On the server console look for message:
	  app Created default stock_setting for enterprise 1! +0ms
  - Verify that the stock settings revert to defaults



Co-authored-by: Jonathan Cameron <jmcameron@gmail.com>

bors[bot] committed:
Merge #4965

4965: chore(deps-dev): bump eslint-plugin-import from 2.22.0 to 2.22.1 r=jniles a=dependabot-preview[bot]

Bumps [eslint-plugin-import](https://github.com/benmosher/eslint-plugin-import) from 2.22.0 to 2.22.1.
<details>
<summary>Changelog</summary>
<p><em>Sourced from <a href="https://github.com/benmosher/eslint-plugin-import/blob/master/CHANGELOG.md">eslint-plugin-import's changelog</a>.</em></p>
<blockquote>
<h2>[2.22.1] - 2020-09-27</h2>
<h3>Fixed</h3>
<ul>
<li>[<code>default</code>]/TypeScript: avoid crash on <code>export =</code> with a MemberExpression (<a href="https://github-redirect.dependabot.com/benmosher/eslint-plugin-import/issues/1841">#1841</a>, thanks [<a href="https://github.com/ljharb">@ljharb</a>])</li>
<li>[<code>extensions</code>]/importType: Fix @/abc being treated as scoped module (<a href="https://github-redirect.dependabot.com/benmosher/eslint-plugin-import/issues/1854">#1854</a>, thanks [<a href="https://github.com/3nuc">@3nuc</a>])</li>
<li>allow using rest operator in named export (<a href="https://github-redirect.dependabot.com/benmosher/eslint-plugin-import/issues/1878">#1878</a>, thanks [<a href="https://github.com/foray1010">@foray1010</a>])</li>
<li>[<code>dynamic-import-chunkname</code>]: allow single quotes to match Webpack support (<a href="https://github-redirect.dependabot.com/benmosher/eslint-plugin-import/issues/1848">#1848</a>, thanks [<a href="https://github.com/straub">@straub</a>])</li>
</ul>
<h3>Changed</h3>
<ul>
<li>[<code>export</code>]: add tests for a name collision with <code>export * from</code> (<a href="https://github-redirect.dependabot.com/benmosher/eslint-plugin-import/issues/1704">#1704</a>, thanks <a href="https://github.com/tomprats">@tomprats</a>)</li>
</ul>
</blockquote>


Co-authored-by: dependabot-preview[bot] <27856297+dependabot-preview[bot]@users.noreply.github.com>

lomamech committed:
Refactor Depot Management Data test

lomamech committed:
Fix and resolve conflict

lomamech committed:
Fix and resolve conflict in migrate file

lomamech committed:
Fix and resolve conflict

lomamech committed:
Fix and resolve conflict

lomamech committed:
Fix and resolve conflict

lomamech committed:
Fix and resolve conflict

lomamech committed:
Fix and sanitaze conflict

jniles commented on client/src/modules/users/UserDepotManagementModal.html:
```diff
@@ -19,14 +19,37 @@ <h4><i> {{ UsersDepotModalCtrl.user.display_name }} </i></h4>
+    </div>
+    
+    <div class="panel panel-default" style="margin-bottom : 0px;">
+      <div class="panel-heading clearfix" style="text-align : left">
```
Do you need the `clearfix` class here?

jniles replied:
Also, instead of using both class and style, you can also use the class `text-left`.

jniles commented on client/src/modules/users/UserDepotManagementModal.html:
```diff
@@ -19,14 +19,37 @@ <h4><i> {{ UsersDepotModalCtrl.user.display_name }} </i></h4>
+      <div style="margin-left: calc({{depot.$$treeLevel}} * 15px)" class="list-unstyled">
+        <div class="checkbox">
+          <label data-label="{{child._label}}">
+            <input type="checkbox" id="depot_{{$index}}" ng-model="depot._checked" ng-change="UsersDepotModalCtrl.setNodeValue(depot.children, depot)" />
```
Instead of `id=depot_{{$index}}`, why don't use use the depot uuid?  That way, it is guaranteed to be unique and you can refer to the depot by its UUID directly.

lomamech replied:
It was only a trick with for the **end to end** test

jniles replied:
As mentioned in person, if you could use an identifier from the depot, rather than it's placement in the array, this will save a lot of test errors in the future.

jniles commented on client/src/modules/users/UserDepotManagementModal.js:
```diff
@@ -19,29 +20,97 @@ function UsersDepotManagementController($state, Users, Notify, AppCache, params)
     vm.user.depots = depots;
   };
 
+  function setNodeValue(childrens, depot) {
```
Hmm... we have a lot of the same functionality in the `bhCheckboxTree`.  I know you tried to use that already, so you don't need to do anything now.

But in a future PR, we should probably add the functionality you are looking for to the `bhCheckboxTree`.  That way, we can test the component to make sure it works correctly.

lomamech replied:
Ok

jniles commented on server/controllers/inventory/depots/index.js:
```diff
@@ -90,7 +105,7 @@ function update(req, res, next) {
           allow_entry_purchase, allow_entry_donation, allow_entry_integration, allow_entry_transfer,
           allow_exit_debtor, allow_exit_service, allow_exit_transfer, allow_exit_loss,
-          min_months_security_stock
+          min_months_security_stock, IF(parent_uuid IS NULL, 0, BUID(parent_uuid)) as parent_uuid
```
Just FYI - there is actually a function for this: [`IFNULL(x, y)`](https://www.w3schools.com/sql/func_mysql_ifnull.asp).

jniles commented on server/models/schema.sql:
```diff
@@ -461,6 +461,7 @@ CREATE TABLE `depot` (
   `allow_exit_loss` TINYINT(1) UNSIGNED NOT NULL DEFAULT 0,
   `location_uuid` BINARY(16) NULL,
   `min_months_security_stock` SMALLINT(5) NOT NULL DEFAULT 2,
+  `parent_uuid` BINARY(16) NULL,
```
Let's add an INDEX on this so allow us to make faster lookups in the future.

jniles reviewed (commented):
@lomamech, this looks pretty good.  I haven't tested the functionality, but made some suggestions to clean up the code.  Most of these comments are non-blocking, but a few could use some touching up.

Once this gets rebased and those comments addressed, we can probably merge it any time.

lomamech committed:
improvement(Depot Management)
- Use sub depot instead dependent depot
- Clear and sanitaze code
- set parent_uuid INDEX

lomamech committed:
Refactor: Use uuid like identifiant for define user depot manangement
